### ISSUE TITLE
Learnable Companding Quantization for Accurate Low-bit Neural Networks

### ISSUE THREAD

**Issue by ryoherisson** [CompML/survey-model-compression#36]
## 一言でいうと
2ビット，3ビット，4ビットモデルの非一様量子化手法を提案．

### 著者/所属機関
Kohei Yamamoto
(Oki Electric Industry Co., Ltd.)

### 投稿日付(yyyy/MM/dd)
2021/03/12

## 概要
量子化関数のうち，非一様量子化は一様量子化よりも量子化誤差や予測誤差を低減することが期待される．しかし，効果的な量子化レベルの正確な推定は容易ではなく，特に低ビットモデルでは一様量子化に比べて精度が劣ることが多い．

本論文では，通信や信号処理の分野で広く使用されるCompanding(Compression and Expansion)技術に基づく，非一様量子化手法のLearnable Companding Quantization(LCQ)を提案．この手法では，入力信号のダイナミックレンジを非線形に制限することでビット幅を縮小する．

これにより，非線形関数とその逆関数を用いることで入出力間のスケールを変化させず，スケールを維持することで量子化誤差を減らし，バックプロパゲーションによる学習を安定化できる．また，Companding関数が微分可能であれば，そのパラメータを最適化することで，タスクの損失を直接最小化できる．

CIFAR-10/100，ImageNet，COCOデータセットで良好な精度を示した．

![スクリーンショット 2021-06-17 8 47 43](https://user-images.githubusercontent.com/42690204/122309136-afdca500-cf48-11eb-8c12-fd7d123ed629.png)

## 新規性・差分
- コンパンド関数を定式化することで，学習によって量子化レベルを柔軟かつ不均一に制御
- 正規化の有効範囲を重み量子化器(LWN)のみに限定することで，量子化学習を安定化できることを発見
- 再量子化を行うことで，効率的な推論に必要なLUTのメモリコストを削減

## 手法
 LCQ quantizer
![スクリーンショット 2021-06-17 10 14 47](https://user-images.githubusercontent.com/42690204/122315357-d99bc900-cf54-11eb-8719-52ba13c4c1f8.png)

Backpropagation for companding
![スクリーンショット 2021-06-17 10 18 49](https://user-images.githubusercontent.com/42690204/122315647-6d6d9500-cf55-11eb-9f7f-67e1903b2cef.png)
![スクリーンショット 2021-06-17 10 19 06](https://user-images.githubusercontent.com/42690204/122315659-73637600-cf55-11eb-9188-fa3f4df96d29.png)

Backpropagation for clipping
![スクリーンショット 2021-06-17 10 19 23](https://user-images.githubusercontent.com/42690204/122315680-7d857480-cf55-11eb-90c5-53a8d4119cf0.png)

Limited Weight Normalization
![スクリーンショット 2021-06-17 10 17 43](https://user-images.githubusercontent.com/42690204/122315553-41eaaa80-cf55-11eb-9b8e-472f8ba3b0d4.png)

## 結果
![スクリーンショット 2021-06-17 10 15 56](https://user-images.githubusercontent.com/42690204/122315436-0223c300-cf55-11eb-85e7-653b5f7f034e.png)
![スクリーンショット 2021-06-17 10 16 02](https://user-images.githubusercontent.com/42690204/122315442-064fe080-cf55-11eb-8e5e-e02d1424697b.png)
![スクリーンショット 2021-06-17 10 16 33](https://user-images.githubusercontent.com/42690204/122315465-18318380-cf55-11eb-8fd9-72c84114402d.png)
